### PR TITLE
Fix: scheduler: prevent resources from running on multiple nodes due to partial live migration

### PR DESCRIPTION
**Pull request by gao-yan**
This consists of multiple fixes/improvements/refactoring in regard of live migration and is mainly to address scenarios originally brought up from:
https://github.com/ClusterLabs/pacemaker/pull/2146#discussion_r837543858

(It's easier to produce with cluster property batch-limit=1):

- rscA is live-migrating from node1 to node2.

- Before migrate_to on node1 returns, put node2 into standby.

- Transition aborts upon return of successful migrate_to on node1. New
  transition is going to stop the rscA on both nodes and start it on
  node1.

- While it is stopping on node1, run something that is going to make
  the transition abort again like:
  crm_resource  --resource rscA --ban --node node2

- Transition aborts upon return of stop on node1.

Now although there's a stop on node1, it's still a partial migration and
rscA is still potentially active on node2. We need to stop it on node2
to prevent it from running on multiple nodes.

The basic idea is to record migrate_to/migrate_from separately from last/last_failure lrm_rsc_op entries, so that the migration history is preserved, which usually matters for multiple nodes and is important for future cluster transitions and should not be overridden.

Then it's to make the scheduler be able to handle the situations when migration history is preserved.

### CODE REVIEW

File: lib/pengine/unpack.c [hunk 1]
```diff
@@ -2205,6 +2205,18 @@ process_rsc_state(pe_resource_t * rsc, pe_node_t * node,
...
+     * the partially migrated resource stopped on the migration target.
+     */
+    if (rsc->role == RSC_ROLE_STOPPED
+        && rsc->partial_migration_source
+        && rsc->partial_migration_source->details == node->details
+        && rsc->partial_migration_target
+        && rsc->running_on) {
+
+        rsc->role = RSC_ROLE_STARTED;
+    }
```
Comment (kgaillot) [inline lib/pengine/unpack.c:2219]: I wonder if native_add_running() should do

    if rsc->role < RSC_ROLE_STARTED rsc->role = RSC_ROLE_STARTED

I can't think of a reason why rsc->role should be RSC_ROLE_STOPPED if rsc->running_on is not NULL

File: cts/scheduler/summary/not-reschedule-unneeded-monitor.summary [hunk 2]
```diff
@@ -11,14 +11,15 @@ Current cluster status:
...
+    * sle12-kvm	(ocf:heartbeat:VirtualDomain):	 FAILED [ kimball castor ]
     * Clone Set: cl_sgdisk [sgdisk]:
       * Started: [ castor kimball ]
 
 Transition Summary:
-  * Recover    sle12-kvm     ( castor -> kimball )
+  * Recover    sle12-kvm     ( kimball )
 
 Executing Cluster Transition:
+  * Resource action: sle12-kvm       stop on kimball
```
Comment (kgaillot) [inline cts/scheduler/summary/not-reschedule-unneeded-monitor.summary:22]: I don't understand, why would we need to stop on kimball when there's already a successful stop as the most recent operation?

**Review by kgaillot — COMMENTED** [review]
Thanks for being so comprehensive

**Review by gao-yan — COMMENTED** [review]
Thanks for review and the good points.

**Review by kgaillot — COMMENTED** [review]
OK, I think I'm following everything now. Just a couple small questions ...

**Pull request merged**